feat: added hook to use chrome pdf generator

maharshivpatel opened the pull request:
Added hooks to use chrome for generating pdf ( instead of wkhtmltopdf).

Actual Implementation is in Print Designer : [print_designer#399](https://github.com/frappe/print_designer/pull/399)

`no-docs`

maharshivpatel committed:
feat(minor): add pdf backend hook

added new_pdf_backend hook to run alternate get_pdf function from other apps e.g. print designer.

maharshivpatel committed:
chore: added print template for new backend.

need to update to handle edge cases.

maharshivpatel committed:
chore: Merge branch 'develop' into add_pdf_backend_hook

maharshivpatel committed:
fix: read_file return base64 encoded string

- added flag to read_file to return base64 encoded string
- fixed pdf_header_footer_chrome to remove unused subst function call
- added new_pdf_backend flag to get_print_format_template function

maharshivpatel committed:
Merge branch 'frappe:develop' into add_pdf_backend_hook

rmehta commented on frappe/__init__.py:
```diff
@@ -2092,6 +2095,10 @@ def get_print(
 	from frappe.utils.pdf import get_pdf
 	from frappe.website.serve import get_response_without_exception_handling
 
+	new_pdf_backend = force_new_backend or frappe.get_cached_value(
```
give this a semantic name like `chrome_pdf_generator`

maharshivpatel committed:
Merge branch 'frappe:develop' into add_pdf_backend_hook

maharshivpatel committed:
fix(minor): remove margin bottom from document status

maharshivpatel committed:
chore: renamed arg name and cleanup

- fixed chrome_pdf_backend arg by making it int and using it to properly override which pdf generator (chrome / wkhtmltopdf) to use.
- renamed new_pdf_backend to chrome_pdf_backend.
- added chrome_pdf_backend arg in local.from_dict to avoid passing it 4-5 functions deep

mahsem commented:
Backport to v15

maharshivpatel committed:
chore: Merge branch 'develop' into add_pdf_backend_hook

maharshivpatel committed:
chore: added chrome_pdf_generator field

instead of making field custom in print_designer added field in Print Format DocType

maharshivpatel committed:
fix: check for chrome_pdf_generator in form_dict first.

first check if chrome_pdf_generator is available in form_dict.
if not available which can happen if function is called from other places then set value from args or print_format.

akhilnarang commented on frappe/__init__.py:
```diff
@@ -2080,6 +2082,7 @@ def get_print(
 	password=None,
 	pdf_options=None,
 	letterhead=None,
+	chrome_pdf_generator=None,
```
Add this to the function docstring

akhilnarang commented on frappe/__init__.py:
```diff
@@ -2111,7 +2126,21 @@ def get_print(
+		return html
+
+	if local.form_dict.chrome_pdf_generator:
+		hook_func = frappe.get_hooks("chrome_pdf_generator")
+		if hook_func:
```
```suggestion
		if hook_func := frappe.get_hooks("chrome_pdf_generator"):
```

akhilnarang reviewed (dismissed):
Seems mostly okay, some minor changes

surajshetty3416 commented on frappe/__init__.py:
```diff
@@ -2111,7 +2126,21 @@ def get_print(
+		return html
+
+	if local.form_dict.chrome_pdf_generator:
+		hook_func = frappe.get_hooks("chrome_pdf_generator")
```
```suggestion
		hook_func = frappe.get_hooks("pdf_generator")
```
Let's call it just `pdf_generator`? Default pdf_generator is "WKHTML2PDF", your will be "CHROME"

surajshetty3416 commented on frappe/printing/doctype/print_format/print_format.json:
```diff
@@ -13,6 +13,7 @@
   "standard",
   "custom_format",
   "disabled",
+  "chrome_pdf_generator",
```
```suggestion
  "pdf_generator",
```

This could be "select" field type, default value should be "WKHTML2PDF"... "CHROME" option should come from your app via property setter.

example: https://github.com/frappe/exotel_integration/blob/e124a6077b58d80d31ac54f2963ba22b81f6791b/exotel_integration/utils/utils.py#L5

surajshetty3416 commented on frappe/utils/pdf.py:
```diff
@@ -34,8 +34,11 @@
 
 def pdf_header_html(soup, head, content, styles, html_id, css):
+	path = "templates/print_formats/pdf_header_footer.html"
+	if frappe.local.form_dict.get("chrome_pdf_generator", False):
+		path = "templates/print_formats/pdf_header_footer_chrome.html"
```
Move this template to print_designer? ATM this format will be dead code unless we install Print Designer

maharshivpatel committed:
fix: make pdf_generator select field

As per review, renamed chrome_pdf_generator to pdf_generator and made it a select field.
- default option available for select is wkhtmltopdf.
- default for new print formats is wkhtmltopdf.
- print designer on install adds a new option ( chrome ) to the select field.
- added patch to update existing print format to wkhtmltopdf.

maharshivpatel committed:
fix: code changes as pdf_generator is now select field

surajshetty3416 commented on frappe/printing/doctype/print_format/patches/sets_wkhtmltopdf_as_default_for_pdf_generator_field.py:
```diff
@@ -0,0 +1,7 @@
+import frappe
+
+
+def execute():
+	"""sets "wkhtmltopdf" as default for pdf_generator field"""
+	for pf in frappe.get_all("Print Format", pluck="name"):
+		frappe.db.set_value("Print Format", pf, "pdf_generator", "wkhtmltopdf")
```
Is this required? Just use fallback value in the code instead?

maharshivpatel replied:
yes, fallback is already available in the code. but, if it is not set then when user opens print format if shows the field empty, so for better clarity/ux. I patched old formats. 

also, someone in future might check the pdf_generator field with assumptions that it will be set.

surajshetty3416 replied:
Okay, at least set `update_modified=False`

maharshivpatel replied:
done

maharshivpatel commented:
@surajshetty3416 let's merge this in develop. we can defer backport to v15

maharshivpatel committed:
fix(minor): don't update modified date in patch

maharshivpatel committed:
Merge branch 'develop' into add_pdf_backend_hook

surajshetty3416 committed:
Merge branch 'develop' into add_pdf_backend_hook

maharshivpatel committed:
Merge branch 'develop' into add_pdf_backend_hook

maharshivpatel committed:
Merge branch 'develop' into add_pdf_backend_hook

surajshetty3416 committed:
Merge branch 'develop' into add_pdf_backend_hook

maharshivpatel committed:
chore: move get_print to print_utils.py

paying my share of taxes #28940

akhilnarang commented:
@mergifyio backport version-15-hotfix version-15

federicocalvo commented:
Hi @maharshivpatel ! is the chrome engine only for print formats generated with print designer?

mohsinalimat commented:
@maharshivpatel @akhilnarang 

https://github.com/frappe/frappe/issues/32162
